README
SkypeAPI version 0.01
=====================

The README is used to introduce the module and provide instructions on
how to install the module, any machine dependencies it may have (for
example C compilers and installed libraries) and any other information
that should be provided before the module is installed.

A README file is required for CPAN modules since CPAN extracts the
README file from a module distribution so that people browsing the
archive can use it get an idea of the modules uses. It is usually a
good idea to provide version information here so that people can
decide whether fixes for the module are worth downloading.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires these other modules and libraries:

  blah blah blah

COPYRIGHT AND LICENCE

Put the correct copyright and licence information here.

Copyright (C) 2008 by A. U. Thor

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.8 or,
at your option, any later version of Perl 5 you may have available.




Changes
Revision history for Perl extension SkypeAPI.

0.01  Fri Aug 15 22:04:16 2008
	- original version; created by h2xs 1.23 with options
		-A SkypeAPI



test.pl
#!perl

use strict;

use lib './lib';

use SkypeAPI;
use Data::Dumper;
use SkypeAPI::Command;


my $skype = SkypeAPI->new();
print " skype available=", $skype->attach , "\n";
my $command = $skype->create_command( { string => "GET USERSTATUS"}  );
print  $skype->send_command($command) , "\n";
$command = $skype->create_command( { string => "SEARCH CHATS"}  );
print $skype->send_command($command) , "\n";

$skype->register_handler(\&handler);



sub handler {
    my $skype = shift;
    my $msg = shift;
    my $command = $self->create_command( { string => "GET USERSTATUS"}  );
    print $self->send_command($command) , "\n";
}


$skype->listen();


t/robot.pl
use strict;

use FindBin qw/$Bin/;
use lib '$Bin/../lib';
use SkypeAPI::Robot;
use Data::Dumper;



SkypeAPI::Robot->new->run();





t/skype.pl
use strict;
use FindBin qw/$Bin/;
use lib '$Bin/../lib';
use SkypeAPI;
use Data::Dumper;

    my $skype = SkypeAPI->new({is_verbose =>0});
    sleep 1;
    
    my $status = $skype->get_command('userstatus');
    print $status, "\n";
    my $version = $skype->get_command('skypeversion');
    print $version, "\n";
    my $currentuserhandle = $skype->get_command('currentuserhandle');
    print $currentuserhandle, "\n";
    my $FULLNAME = $skype->get_command('profile fullname');
    print $FULLNAME, "\n";
    
    my $achats = $skype->search_chats('ACTIVECHATS');
    print Dumper($achats), "\n";
    
    my $chats = $skype->search_chats('chats');
    print Dumper($chats), "\n";
    
    my $missedchats = $skype->search_chats('MISSEDCHATS');
    print Dumper($missedchats), "\n";
    
    my $RECENTCHATS = $skype->search_chats('RECENTCHATS');
    print Dumper($RECENTCHATS), "\n";
    
    my $BOOKMARKEDCHATS = $skype->search_chats('BOOKMARKEDCHATS');
    print Dumper($BOOKMARKEDCHATS), "\n";
    
    
    #send message to recenly chat
    if (@$RECENTCHATS > 0) {
        $skype->send_chat_message($RECENTCHATS->[0], '您好,我是帅哥');
    }
   
    my $i1 = $skype->add_message_listener(\&mesasge_listener, 1);
    $skype->message_listen();
    



sub mesasge_listener {
    my ($instance, $message_id) = @_;
    
    print "I received message $message_id\n";
     my $CHATNAME  = $instance->get_message($message_id, 'CHATNAME');
    print "CHATNAME :$CHATNAME \n";
    
    my $body = $instance->get_message($message_id, 'body');
    print "body:$body\n";
    
    return 1;
}


__END__
    my $skype = SkypeAPI->new({is_verbose => 1});
    
    while (my $line = <STDIN>) {
        chomp($line);
        $skype->do_command($line);
        sleep 1;
    }
    
__END__
    my $skype = SkypeAPI->new({is_verbose => 1});
    sleep 5;
    while (my $line = <STDIN>) {
        chomp($line);

        my $cmd_id = $skype->do_command_for_result($line);
        for (1..100) {
            my $result = $skype->pop_command_result($cmd_id);
            print "[$cmd_id]$result", "\n";
            if ($result ne '-1' and $result ne '-2') {
                last;
            }
            sleep 0.1;
        }
        
    }    
    
__END__
    while (my $line = <STDIN>) {
        chomp($line);

        my $cmd_id = $skype->do_command_for_result($line);
        print "$cmd_id=>\n";
        my $result = $skype->wait_message($cmd_id);
        print "$result", "\n";        
    }      

Makefile.PL
use 5.008005;
use ExtUtils::MakeMaker;
# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.
WriteMakefile(
    NAME              => 'SkypeAPI::Win',
    VERSION_FROM      => 'lib/SkypeAPI.pm', # finds $VERSION
    PREREQ_PM         => {}, # e.g., Module::Name => 1.1
    ($] >= 5.005 ?     ## Add these new keywords supported since 5.005
      (ABSTRACT_FROM  => 'lib/SkypeAPI.pm', # retrieve abstract from module
       AUTHOR         => 'laomoi <laomoi@gmail.com>') : ()),
    LIBS              => [''], # e.g., '-lm'
    DEFINE            => '', # e.g., '-DHAVE_SOMETHING'
    INC               => '-I.', # e.g., '-I. -I/usr/include/other'
	# Un-comment this if you add C files to link with later:
    # OBJECT            => '$(O_FILES)', # link all the C files too
);


lib/SkypeAPI/Win.pm
package SkypeAPI::Win;

use 5.008005;
use strict;
use warnings;

require Exporter;
use Class::Accessor::Fast;

our @ISA = qw(Exporter Class::Accessor::Fast);

our $VERSION = '0.07';

require XSLoader;
XSLoader::load('SkypeAPI::Win', $VERSION);

# Preloaded methods go here.
use threads;
use threads::shared;
use Data::Dumper;

__PACKAGE__->mk_accessors(
  qw/thread   is_running/
);


sub init {
    my $self = shift;
    my $handler_list = shift;
    my $thread = new threads(\&run, $self, $handler_list);
    $self->thread($thread);
}

sub run {
    my $self = shift;
    $self->attach( { copy_data => \&handler } );
}

sub handler {
    my $message = shift;
    print "[api]$message\n";
    if (defined $message and $message =~ m{^#(\w+)\s+(.*)}) {
        my ($id, $reply) = ($1, $2);
        if ($SkypeAPI::command_list{$id}) {
            my $command;
            {
                lock %SkypeAPI::command_list;
                $command = delete $SkypeAPI::command_list{$id};
                
            }
            
            {
                lock $SkypeAPI::command_lock;
                
                $command->reply($reply); 
               
            }
            
        }  
    } else {
        lock @SkypeAPI::message_list;
        push @SkypeAPI::message_list, $message;
    }

    #print Dumper(\%SkypeAPI::command_list);
}



sub DESTROY {
    my $self = shift;
    if ($self->thread) {
        $self->quit();
        $self->thread->join;
        $self->thread(undef);
    }
}




1;
__END__


lib/SkypeAPI/Robot.pm
package SkypeAPI::Robot;

use strict;
use warnings;

use SkypeAPI;
use XiaoI;
use Data::Dumper;
our $VERSION = '0.02';

my $instance = undef;

sub new {
    my $class = shift;
    my $opt = shift;
    
    return $instance if $instance;
    
    $instance = {opt => $opt};
    bless $instance, $class;
    $instance->{robot_list} = {};
    

    return $instance;
}

sub mesasge_listener {
    my ($skype, $message_id) = @_;
    
    print "I received message $message_id\n";
     my $CHATNAME  = $skype->get_message($message_id, 'CHATNAME');
    print "CHATNAME :$CHATNAME \n";
    
    if (not exists $instance->{robot_list}->{ $CHATNAME }) {
        print "CREAET NEW ROBOT FOR THE CHAT\n";
        my $robot = XiaoI->new;
        $instance->{robot_list}->{$CHATNAME} = $robot;
    }
    
    my $body = $skype->get_message($message_id, 'body');
    print "body:$body\n";    
   
   
    my $robot = $instance->{robot_list}->{$CHATNAME};
    my $text = $robot->get_robot_text($body);
    $skype->send_chat_message($CHATNAME, $text);
    
    return 1;
}



sub run {
    my $self = shift;   
    
    my $skype = SkypeAPI->new({is_verbose =>0});
    sleep 1;
    
    my $listener = $skype->add_message_listener(\&mesasge_listener, 1);
    $skype->message_listen();
}


1;


lib/SkypeAPI/Command.pm
package SkypeAPI::Command;

use strict;
use warnings;

use Class::Accessor::Fast;
our @ISA = qw(Class::Accessor::Fast);

our $VERSION = '0.01';


# Preloaded methods go here.
use Time::HiRes qw( sleep );

__PACKAGE__->mk_accessors(
  qw/blocking id timeout string reply/
);


1;
__END__
# Below is stub documentation for your module. You'd better edit it!

=head1 NAME

SkypeAPI - Perl extension for blah blah blah

=head1 SYNOPSIS

  use SkypeAPI;
  blah blah blah

=head1 DESCRIPTION

Stub documentation for SkypeAPI, created by h2xs. It looks like the
author of the extension was negligent enough to leave the stub
unedited.

Blah blah blah.

=head2 EXPORT

None by default.



=head1 SEE ALSO

Mention other useful documentation such as the documentation of
related modules or operating system documentation (such as man pages
in UNIX), or any relevant external documentation such as RFCs or
standards.

If you have a mailing list set up for your module, mention it here.

If you have a web site set up for your module, mention it here.

=head1 AUTHOR

A. U. Thor, E<lt>a.u.thor@a.galaxy.far.far.awayE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2008 by A. U. Thor

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.7 or,
at your option, any later version of Perl 5 you may have available.


=cut


lib/SkypeAPI.pm
package SkypeAPI;

use 5.008005;
use strict;
use warnings;

require Exporter;
use Class::Accessor::Fast;
our @ISA = qw(Exporter Class::Accessor::Fast);

our $VERSION = '0.07';


# Preloaded methods go here.
use SkypeAPI::Win;
use Time::HiRes qw( sleep );
use threads::shared;
use Digest::MD5;
use Data::Dumper;

$| = 1;

__PACKAGE__->mk_accessors(
  qw/api handler_list stop_listen/
);

our $command_lock : shared;
our %command_list : shared;
our @message_list : shared;

sub attach {
    my $self = shift;
    print "[skype]I am attaching\n";    
    
    $self->api(SkypeAPI::Win->new);
    $self->api->init($self->handler_list);
    $self->wait_available;
}

sub is_available {
    my $self = shift;
    return $self->api->is_available();
}

sub wait_available {
    my $self = shift;
    my ($times, $interval) = @_;
    $times ||= 100;
    $interval ||= 0.1;
    for (1..$times) {
        return 1 if $self->api->is_available();
        sleep $interval;
    }
    return 0;
}

sub send_command {
    my $self = shift;
    my $command = shift;
    
    {
        lock(%command_list);
        $command_list{$command->id} = $command;
    }
    
    
    my $command_string = sprintf("#%s %s", $command->id, $command->string);
    printf("[send]$command_string\n");
    $self->api->send_message($command_string);
    
    if ($command->blocking) {
        {
            my $start = time;
            WAIT:
            while (1) {
                {
                    lock $SkypeAPI::command_lock;
                    if (defined $command->reply) {
                        last WAIT;
                    }
                }                
                
                if (time - $start > $command->timeout) {
                    print "[skype]wait command timeout\n";
                    last;
                }
                sleep 0.01;
            }           
            
        }
    }
    return $command->reply();
}

sub create_command {
    my $self = shift;
    my $opt = shift;
    my $command;
    share($command);
    $command = &share({});
    bless $command, 'SkypeAPI::Command';
    $command->timeout($opt->{timeout} || 10);
    $command->blocking($opt->{blocking} || 1);
    $command->id($opt->{id} || substr(Digest::MD5::md5_hex(time . int(rand(10000))), 0, 16));
    $command->string($opt->{string} || '');
    return $command;
}

sub register_handler {
    my $self = shift;
    my $ref_sub = shift;
    $self->handler_list({}) if not defined $self->handler_list;
    my $id = keys %{$self->handler_list};
    $self->handler_list->{$id + 1} = $ref_sub;    
}

sub listen {
    my $self = shift;

    $self->stop_listen(0);
    while (!$self->stop_listen) {
        my $message;
        {
            lock @message_list;
            $message = shift @message_list;
        }
        if (not defined $message) {
            sleep 0.01;             
            next;
        }
        for my $id (sort keys %{$self->handler_list}) {
            $self->handler_list->{$id}->($self, $message);
        }
    }
}



1;
__END__

=head1 NAME

SkypeAPI - Skype API simple implementation, only support windows platform now.

=head1 VERSION

0.07

=head1 SYNOPSIS

    use SkypeAPI;
    my $skype = SkypeAPI->new();
    print " skype available=", $skype->attach , "\n";
    my $command = $skype->create_command( { string => "GET USERSTATUS"}  );
    print  $skype->send_command($command) , "\n";
    $command = $skype->create_command( { string => "SEARCH CHATS"}  );
    print $skype->send_command($command) , "\n";

=head1 FUNCTIONS

=head2 SkypeAPI->new( )

Returns a SkypeAPI object. 

    my $skype = SkypeAPI->new();

=head2 SkypeAPI->attach( )

Attach to skype, return 1 if attached ok.

=head2 SkypeAPI->send_command( $command_object )

Send command to skype, return the reponse of skype

=head2 SkypeAPI->create_command( $opt )

Create command. $opt is a hashref. You can defind the command string, timeout in it.

timeout default is 10 seconds.

    my $command = $skype->create_command( { string => "SEARCH CHATS", timeout => 5}  );


=head2 SkypeAPI->register_handler( $ref_callback )

Add listener to the chain of message handler. 

    $skype->register_handler(\&handler);
    $skype->attach();
    sub handler {
        my $skype = shift;
        my $msg = shift;
        my $command = $skype->create_command( { string => "GET USERSTATUS"}  );
        print $skype->send_command($command) , "\n";
    }

=head2 SkypeAPI->listen( )

After you register handlers, call $skype->listen to enter the message loop;

    $skype->listen();

=head2 SkypeAPI->stop_listen( )

Call stop_listen in your handler to exit the message loop

    sub handler {
        my $skype = shift;
        my $msg = shift;
        $skype->stop_listen();
    }



=head1 DESCRIPTION

A Perl simple implementation of the Skype API, working off of the canonical Java and Python implementations.
It is a encapsulation of Windows message communication between Skype and client applications.
 This version of SkypeAPI only implement some commands of SKYPE API, you can implement the others using  SkypAPI->send_command

=head2 EXPORT

None by default.

=head1 ROBOT DEMO

You can find the robot.pl in the lib/../t/robot.pl, run it and your skype will become a xiaoi robot :)

the robot needs the module XiaoI, please install it first, See L<http://code.google.com/p/xiaoi/>


=head1 SEE ALSO

For more command information, See L<https://developer.skype.com/Docs/ApiDoc/src>

The svn source of this project, See L<http://code.google.com/p/skype4perl/>


=head1 AUTHOR

laomoi ( I<laomoi@gmail.com> )

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2008 by laomoi

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itinstance, either Perl version 5.8.8 or,
at your option, any later version of Perl 5 you may have available.


=cut
